### ISSUE TITLE
Help getting started the right way

### ISSUE THREAD

**Issue by theArianit** [tj/n#725]
# Installation and setup guide for absolute noobs

## Motivation
I am using opensuse tumbleweed as my OS (I am a complete beginner in linux). I installed nodejs with npm from the official tumbleweed repo. After installing a global package I got the message to update npm to a newer version. Little did I know that the versions from the repo and official releases don't always correspond. So I updated npm from the CLI, which caused problem with the directories

## Guides and help beyond the readme
After asking opensuse's reddit and learning about the issued as described in my motivation, I got suggested to use a node version manager. Most of the others I looked on required to do some bash configuration, profile settings, etc. which for me as a beginner were scary. So n seems to go straight to the magic without all the ingredients and sacrifices.
But I still have some questions and things I don't understand, and I don't want to do anything without understanding a bit before doing. So, be patience with my ignorance.

- As much as I understood, n will replace the installed npm and node versions (node version with the versions I install with n). Or how does this exactly work? Or to get to the point, what will happen with the nodejs and npm version that got installed from the opensuse distro after I install n and install other nodejs? What if I install the same nodejs version as installed from opensuse?
- I usually like to have the latest versions of apps, So I would like to have the latest npm version (without causing other problems). How to?
- `N_PREFIX`. What exactly will this do, will it only change the directory where n will be installed or also the directory where global npm packages get installed? Please give me an example how to use this command on installation.
- Is it safe to install global packages with n as node version manager?
- (This is maybe not specific to n) How to make global packages be installed into my home directory instead of the standard directory (_/usr/local/lib/node_modules_) and does this make sense? I want them to be under something like _/home/arianit/.npm-globalls/_, where I intend to install mostly cli's.

I hope anyone has the patience to help me and explain these things in layman terms.

**Comment by shadowspawn** [comment]
Some links to previous related discussions:
- https://github.com/tj/n/issues/709#issuecomment-1014148143 permissions and install location
- https://github.com/tj/n/issues/709#issuecomment-1014154924 `PATH`
- #716 using latest npm version

**Comment by shadowspawn** [comment]
> As much as I understood, n will replace the installed npm and node versions (node version with the versions I install with n). 

This depends on where the old and new versions get installed. 

If the new and old are installed to the same location as happens on macOS with default install location of `/usr/local` then the new version overwrites the old version.

If the new and old are installed to different locations, whether because using `N_PREFIX` or because opensuse distro uses a different location, then you have two versions on disk. Which one gets launched when you type `node` depends on what is in your `PATH`. Unless you particularly want to keep the distro version, it can reduce confusion to delete that one.

**Comment by shadowspawn** [comment]
> `N_PREFIX`. What exactly will this do, will it only change the directory where n will be installed or also the directory where global npm packages get installed?

By default the global npm packages are installed relative to the npm location. So implicitly using `N_PREFIX` does also change where global npm packages are installed.

**Comment by shadowspawn** [comment]
> Is it safe to install global packages with n as node version manager?

Yes. (For example, a suggested way of installing `n` is as an npm global package!)

**Comment by theArianit** [comment]
So this is what I get out from all of this:

- nodejs and npm installed from the OS get overridden by the ones installed via `n` if I configure them to be installed in the same location. Here I get another question. What if I delete those installed from the OS after I install `n`? This is the part where I have the most confusion. According to your instructions from the first link from your first reply, to change the location of globally installed npm packages I should use the command `npm set prefix ~/.npm-global`, this was point 4 from your instructions. But I don't get the third point. 

> 3. Use a node version manager, and install and run node and npm out of your home folder. Then when npm installs the global packages "beside" itself, it is in your home folder. Changing the Node.js install location is what N_PREFIX is for, and what n-install does for you.

Can you please break it up for me.

- `n --preserve` will keep the latest installed npm version even when changing node versions. This is now clear.
- Setting up the path. `export PATH=/Users/nanimans/.npm-global/bin:$PATH` this will tell the bash shell where to look for the executables. For example when I install angular globally and run `ng new demoproj`, bash will look inside .npm-globals for the npm packages and executables for angular. Correct me if I am wrong. 
- `export N_PREFIX=$HOME/.n` and `export PATH=$N_PREFIX/bin:$PATH` will configure where the node versions get installed and where to look for them. Right? But then, what's the difference between `N_PREFIX` vs `npm set prefix`?

- Your last comment: 

> Yes. (For example, a suggested way of installing n is as an npm global package!)

I think you misunderstood me. What I mean is that after I install `n` and set up everything, and then I decide to install a global package, will that global installation have any side effects (because of different node versions and so)? example `npm install -g @angular/cli`, will this cause any problems or is it safe to install global npm packages after `n` is installed on my machine?

I have the feeling that these are already answered but there is some barrier (language mostly) which doesn't let me grasp it 100%.
What will I lose/gain if I leave everything on default?
For your information, `npm root -g` says `/usr/local/lib/node_modules`. And `which node` says '/usr/bin/node' and `which npm` says `/usr/bin/npm`

**Comment by shadowspawn** [comment]
I got one detail wrong in above comments, and have corrected the comment.

By default the global npm packages are installed relative to the node location. Reference: https://docs.npmjs.com/cli/v8/configuring-npm/folders

**Comment by shadowspawn** [comment]
> What I mean is that after I install n and set up everything, and then I decide to install a global package, will that global installation have any side effects (because of different node versions and so)?

By default, npm installs global packages relative to the node executable.

Per the README:

> `n` downloads a prebuilt Node.js package and installs to a single prefix (e.g. `/usr/local`). This overwrites the previous version.

Changing Node.js versions using `n` will not change what global packages are installed and available (other than packages which are part of Node.js like npm). Not sure if you count that as a yes or a no for side effects!

(Technical note: this behaviour is different to some other node version managers which install the global packages separately per Node.js version. They update the PATH or a symlink or search for the appropriate node version to launch.)

**Comment by shadowspawn** [comment]
> What if I delete those [node and npm] installed from the OS after I install n? 

There are a host of assumptions in that question, and I'll try and clarify a few things and let you try asking again.

Installing `n` does not change what node or npm you have installed. That happens when you use `n` to install Node.j, like `n install lts`.

-----

When you install a global package using npm, the default location is relative to the node executable. If node is installed in a "system" folder like `/usr/local/bin/node` then the global packages will also be written to a system folder. It is a bad idea to use sudo with all your npm global installs, so either you can take ownership of the necessary system folders, or you can change where node is installed to under your home folder, or you can change where npm global packages are installed by configuring npm.

I suggest configuring the global npm prefix is an extra complication and reasonably rare. Hence if you want things under your home folder, use `N_PREFIX` so node is installed under your home folder, and the node global packages will be too.

-----

> `export N_PREFIX=$HOME/.n` and `export PATH=$N_PREFIX/bin:$PATH` will configure where the node versions get installed and where to look for them. Right? But then, what's the difference between `N_PREFIX` vs `npm set prefix`?

Right. And when you install a global npm package, npm puts a symlink for any binaries into that bin folder too, beside node. So you don't need an extra entry in the PATH to find the binaries belonging to the global packages, they are covered by the same entry that finds node.

`npm set prefix` is changing where the global packages get installed which is _separate_ from accessing the binaries associated with the global packages.

**Comment by shadowspawn** [comment]
> What will I lose/gain if I leave everything on default?

You want a newly installed version of node by `n` to be in your `PATH`. The default location for node is `/usr/local/bin/node`. Is `/usr/local/bin` in your PATH? If so, then leaving things on default avoids needing to alter `PATH`.

If you leave everything on default, instructions off the internet are more likely to work.

To avoid using `sudo` with npm global installs, you can change the permissions of the necessary system folders. You might prefer not to though, or be unable to, in which case installing to your home folder solves the permission issue but requires modifying `PATH`.

**Comment by shadowspawn** [comment]
You originally commented:

> So I updated npm from the CLI, which caused problem with the directories

What problems with the directories? Lack of permission to write to a system folder? This is the same issue I keep mentioning, and the options are changing permission or installing to somewhere else.

**Comment by shadowspawn** [comment]
> `N_PREFIX`. What exactly will this do, will it only change the directory where n will be installed or also the directory where global npm packages get installed? Please give me an example how to use this command on installation.

`N_PREFIX` changes where `n` installs Node.js when you run installs. Since this changes where node gets installed, it also changes where npm global packages get installed when running that version of node.

You put the definition for `N_PREFIX` in your shell startup file (same file you use to make changes to `PATH`).

**Comment by theArianit** [comment]
Thank you for clarifying things for me. So this `N_PREFIX` and `n --preserve` seems to be all I need to do. But, after I do this, and install a node version from `n`, is it then safe to delete the nodejs that I installed from the OS or should that always be on my machine as long as there is `n` installed and I want to use it?

> What problems with the directories? Lack of permission to write to a system folder? This is the same issue I keep mentioning, and the options are changing permission or installing to somewhere else.

The directories changed. So instead of looking in /usr/local/lib it did looked in /usr/lib and I couldn't use npm anymore because it always said that directory or command could not be found.

**Comment by shadowspawn** [comment]
> But, after I do this, and install a node version from `n`, is it then safe to delete the nodejs that I installed from the OS or should that always be on my machine as long as there is `n` installed and I want to use it?

`n` is actually a bash script, so it is ok to delete the distro node and `n` will keep working. `n` does not need node installed to work.

> The directories changed. 

Oh, that might be specific to the distro setup. But it raises a good issue.

When you change where node is installed and are using default behaviours, you implicitly change where npm installs new global packages. So before making changes to where node is installed, you might want to list what you currently have installed in case you need to reinstall them in the new preferred location (and/or delete them from the old location).

```
npm list --global --depth=0
```

**Comment by shadowspawn** [comment]
>  So this N_PREFIX and n --preserve seems to be all I need to do.

If you are usually going to preserve npm, you can make this the default rather than type it each time. See `N_PRESERVE_NPM`.

**Comment by theArianit** [comment]
> If you are usually going to preserve npm, you can make this the default rather than type it each time. See N_PRESERVE_NPM.

I have to type it each time? when do I have to do so?

**Comment by shadowspawn** [comment]
Define the environment variable in the initialization file for your shell, the same place as `N_PREFIX`. Then start a new shell.

**Comment by shadowspawn** [comment]
> I have to type it each time?

By default each time you use `n` to install a version of Node.js it will overwrite npm since that is part of the install. 

So if you intend to normally preserve the (presumably current) version of npm when you switch node versions, you need to add `-p` or `--preserve`. Or you might define `N_PRESERVE_NPM` to make the default behaviour that npm is preserved, and opt-out when you occasionally want to install the npm that came with that version of Node.js.

**Comment by theArianit** [comment]
Sorry for being ignorant but I want to keep it as simple as possible, where I do things once and don't worry anymore. Because it feels like just when I think I get it, then something confuses me and it's all from the beginning.
I will describe the steps and configs I will make to make things work for my scenario, and please correct me for things I get wrong
(by order of execution)
1. So, I install `n`, set the `export N_PREFIX=$HOME/.n`. Locations of node installation and executables.
2. Delete the node and npm installed from my OS. (because I don't want to deal with them anymore).
3. Set up the shell path `export PATH=~/.npm-global/bin:$PATH`. Location where to look for globally installed npm packages and executables.
4. Install node js (and npm) via `n`.
5. Set the location where to store global npm packages `npm set prefix ~/.npm-global`. Step 3 will look here for executables.
6. Preserve npm by default `export N_PRESERVE_NPM=1`.

**Comment by shadowspawn** [comment]
You are missing a PATH entry for where the Node.js binaries get installed. From the README:

```
export N_PREFIX=$HOME/.n
export PATH=$N_PREFIX/bin:$PATH
```

**Comment by theArianit** [comment]
> You are missing a PATH entry for where the Node.js binaries get installed

Right, thank you.

**Comment by theArianit** [comment]
So, I have followed all the steps from my recent comment. I installed one global package (@tsed/cli) and when I want to init a project with it (`tsed init .`), it says "bash: tsed: command not found".
What may be the cause and how can I check that the exports made the right changes?

**Comment by shadowspawn** [comment]
That error message means there is not an executable called `tsed` in your `PATH`.

You can see all the environment variables by running `env`, or a single variable using `echo $FOO`.

To diagnose what may be the cause check
1) where is npm writing global modules?
2) is there an executable called `tsed ` in the `bin` folder there?
3) is that `bin` folder in your PATH? The folders in `PATH` are separated by `:`.

**Comment by theArianit** [comment]
I think the 3rd option is my error case. Is the PATH in this .bashrc file?

**Comment by shadowspawn** [comment]
Definitions for `N_PREFIX ` and `PATH` and `N_PRESERVE_NPM` will all go in the same startup file. 

The appropriate file varies depending on what shell you are using. I don't offer advice on working that out, sorry!

**Comment by theArianit** [comment]
I did 
`export N_PREFIX=$HOME/.n
export PATH=$N_PREFIX/bin:$PATH
export PATH=~/.npm-global/bin:$PATH`
I don't know what's wrong neither how to check if those changes were applied

**Comment by theArianit** [comment]
I added those export manually in the ~/.bashrc files and it works for now
